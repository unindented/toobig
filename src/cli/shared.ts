import { cosmiconfigSync } from "cosmiconfig";
import { existsSync, lstatSync } from "fs-extra";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const pkg = require("../../package.json") as { name: string };

export const parserConfiguration = {
  "boolean-negation": true,
  "camel-case-expansion": false,
  "combine-arrays": false,
  "dot-notation": false,
  "duplicate-arguments-array": true,
  "flatten-duplicate-arrays": false,
  "halt-at-non-option": false,
  "negation-prefix": "no-",
  "parse-numbers": false,
  "populate--": false,
  "set-placeholder-key": false,
  "short-option-groups": false,
  "strip-aliased": false,
  "strip-dashed": false,
  "unknown-options-as-args": false,
};

export const configParser = (configPath?: string): object => {
  const explorer = cosmiconfigSync(pkg.name);
  const result = isFile(configPath)
    ? explorer.load(configPath)
    : explorer.search(configPath);

  return result != null && typeof result.config === "object"
    ? (result.config as object)
    : {};
};

const isFile = (path?: string): path is string =>
  !!(path && existsSync(path) && lstatSync(path).isFile());
