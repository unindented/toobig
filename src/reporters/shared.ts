import { basename, dirname, relative, resolve as resolvePath, sep } from "path";

import bytes from "bytes";
import terminalLink from "terminal-link";

import {
  getBaselineDifference,
  getTotalBaselinesDifference,
  getTotalBaselinesSize,
  getTotalResultsSize,
  isOverBudget,
} from "../shared";
import {
  ResultContext,
  ResultWithBaselineContext,
  ResultsContext,
  ResultsWithBaselinesContext,
} from "../types";

export const formatSize = ({
  result,
}: Partial<Pick<ResultContext, "result">>): string =>
  result ? bytes(result.size) : "";

export const formatMaxSize = ({
  result,
}: Partial<Pick<ResultContext, "result">>): string =>
  result ? bytes(result.maxSize) : "";

export const formatBaselineSize = ({
  baseline,
}: Partial<Pick<ResultWithBaselineContext, "baseline">>): string =>
  baseline ? bytes(baseline.size) : "";

export const formatSizeVsMaxSize = (context: ResultContext): string => {
  const {
    result,
    outputContext: { colors, colorOverride },
  } = context;
  const overBudget = isOverBudget(result);
  const color = colorOverride
    ? colorOverride
    : overBudget
    ? colors.bold.red
    : colors.dim;
  const operator = overBudget ? ">=" : "<";

  return color(`${formatSize(context)} ${operator} ${formatMaxSize(context)}`);
};

export const formatSizeVsBaselineSize = (context: ResultContext): string => {
  const {
    outputContext: { colors, colorOverride },
  } = context;
  const overBaseline = getBaselineDifference(context) > 0;
  const color = colorOverride
    ? colorOverride
    : overBaseline
    ? colors.bold.red
    : colors.dim;
  const operator = overBaseline ? ">=" : "<";

  return color(
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    `${formatSize(context)} ${operator} ${
      formatBaselineSize(context) || bytes(0)
    }`
  );
};

export const formatBaselineDifference = (context: ResultContext): string => {
  const difference = getBaselineDifference(context);
  return difference !== 0 ? bytes(difference) : "";
};

export const formatTotalResultsSize = (context: ResultsContext): string =>
  bytes(getTotalResultsSize(context));

export const formatTotalBaselinesSize = (
  context: ResultsWithBaselinesContext
): string => bytes(getTotalBaselinesSize(context));

export const formatTotalBaselinesDifference = (
  context: ResultsWithBaselinesContext
): string => {
  const difference = getTotalBaselinesDifference(context);
  return difference !== 0 ? bytes(difference) : "";
};

const failMark = "\u2717";
const passMark = "\u2713";

export const formatMaxSizeMark = ({
  result,
  outputContext: { colors },
}: ResultContext): string =>
  isOverBudget(result) ? colors.red(failMark) : colors.green(passMark);

const upArrow = "\u25B2";
const downArrow = "\u25BC";

export const formatBaselineDifferenceArrow = (
  context: ResultWithBaselineContext
): string => {
  const {
    outputContext: { colors },
  } = context;
  const difference = getBaselineDifference(context);
  return difference > 0
    ? colors.red(upArrow)
    : difference < 0
    ? colors.green(downArrow)
    : " ";
};

export const formatTotalBaselinesDifferenceArrow = (
  context: ResultsWithBaselinesContext
): string => {
  const {
    outputContext: { colors },
  } = context;
  const difference = getTotalBaselinesDifference(context);
  return difference > 0
    ? colors.red(upArrow)
    : difference < 0
    ? colors.green(downArrow)
    : " ";
};

const ellipsis = "\u2026";

const truncateString = (str: string, maxLength: number): string =>
  str.length <= maxLength ? str : `${str.substr(0, maxLength - 1)}${ellipsis}`;

export const formatPath = ({
  result: { path },
  outputContext: { colors, maxLength },
}: ResultContext): string => {
  const relativePath = relative(process.cwd(), path);
  const dir = dirname(relativePath);
  const dirWithSep = dir !== "." ? dir + sep : "";
  const base = truncateString(
    basename(relativePath),
    maxLength - dirWithSep.length
  );
  return `${colors.dim(dirWithSep)}${colors.bold(base)}`;
};

export const linkPath = (resultContext: ResultContext): string => {
  const formattedPath = formatPath(resultContext);
  return terminalLink(
    formattedPath,
    `file://${resolvePath(process.cwd(), resultContext.result.path)}`,
    {
      fallback: () => formattedPath,
    }
  );
};
