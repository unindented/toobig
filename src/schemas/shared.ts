import Joi from "@hapi/joi";

export const reportersConfigSchema = Joi.array().items(
  Joi.alternatives().try(
    Joi.string(),
    Joi.array().items(Joi.string().required(), Joi.object().required())
  )
);

export const resultsConfigSchema = Joi.array().items(
  Joi.object({
    path: Joi.string().required(),
    size: Joi.number().required(),
    maxSize: Joi.number().required(),
  }).required()
);
