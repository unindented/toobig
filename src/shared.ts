import { parse } from "url";

import axios from "axios";
import {
  Instance as ChalkInstance,
  supportsColor as globalSupportsColor,
} from "chalk";
import { createReadStream, createWriteStream, ensureFileSync } from "fs-extra";

import {
  InputStream,
  OutputContext,
  OutputStream,
  Result,
  ResultContext,
  Results,
  ResultsContext,
  ResultsWithBaselinesContext,
  WritableResults,
} from "./types";

export { globalSupportsColor as supportsColor };

export const isOverBudget = ({ size, maxSize }: Result): boolean =>
  size >= maxSize;

export const getBaselineDifference = ({
  result,
  baseline,
}: Pick<ResultContext, "result" | "baseline">): number =>
  result.size - (baseline ? baseline.size : 0);

export const isUnderBaseline = (baselines?: Results) => (
  result: Result
): boolean => {
  if (baselines === undefined) {
    return false;
  }

  const baseline = baselines[result.path];
  return getBaselineDifference({ result, baseline }) < 0;
};

export const isOverBaseline = (baselines?: Results) => (
  result: Result
): boolean => {
  if (baselines === undefined) {
    return false;
  }

  const baseline = baselines[result.path];
  return getBaselineDifference({ result, baseline }) > 0;
};

export const filterResults = (
  results: Results,
  predicate: (result: Result) => boolean
): Results =>
  Object.keys(results).reduce((acc, key) => {
    if (predicate(results[key])) {
      acc[key] = results[key];
    }
    return acc;
  }, {} as WritableResults);

export const getTotalResultsSize = ({
  results,
}: Pick<ResultsContext, "results">): number => getTotalSize(results);

export const getTotalBaselinesSize = ({
  baselines,
}: Pick<ResultsWithBaselinesContext, "baselines">): number =>
  getTotalSize(baselines);

const getTotalSize = (results: Results): number =>
  Object.values(results).reduce((acc, { size }) => acc + size, 0);

export const getTotalBaselinesDifference = ({
  results,
  baselines,
}: Pick<ResultsWithBaselinesContext, "results" | "baselines">): number =>
  Object.values(results).reduce((acc, { path, size }) => {
    const baseline = baselines[path];
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    return acc + size - (baseline ? baseline.size : 0);
  }, 0);

export const getOutputContext = ({
  supportsColor,
  maxLength = Infinity,
}: {
  supportsColor: boolean;
  maxLength?: number;
}): OutputContext => ({
  colors: new ChalkInstance(!supportsColor ? { level: 0 } : undefined),
  maxLength,
});

export const getOutputStream = (
  output: string | NodeJS.WritableStream
): OutputStream => {
  if (output === "stdout" || output === "stderr") {
    return process[output];
  }

  if (typeof output === "string") {
    ensureFileSync(output);
    return createWriteStream(output);
  }

  return output;
};

export const endOutputStream = (stream: OutputStream): void => {
  if (stream === process.stdout || stream === process.stderr) {
    return;
  }

  stream.end();
};

export const getInputStream = async (
  input: string | NodeJS.ReadableStream
): Promise<InputStream> => {
  if (input === "stdin") {
    return process.stdin;
  }

  if (typeof input === "string") {
    const { protocol } = parse(input);
    return protocol === null
      ? createReadStream(input)
      : createAxiosStream(input);
  }

  return input;
};

const createAxiosStream = async (
  input: string
): Promise<NodeJS.ReadableStream> => {
  const { data: stream } = await axios.get<NodeJS.ReadableStream>(input, {
    responseType: "stream",
    timeout: 10000,
  });
  return stream;
};

export const readInputStream = (stream: InputStream): Promise<string> => {
  return new Promise((resolve, reject) => {
    const contents: string[] = [];

    stream.on("data", (data: string | Buffer) => {
      contents.push(data.toString());
    });

    stream.on("end", () => {
      resolve(contents.join(""));
    });

    stream.on("error", (error: Error) => {
      reject(error);
    });
  });
};
