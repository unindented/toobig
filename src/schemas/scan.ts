import Joi from "@hapi/joi";

import { reportersConfigSchema } from "./shared";

export const scanConfigSchema = Joi.object({
  budgets: Joi.object().pattern(/.+/, Joi.string()).required(),
  cwd: Joi.string(),
  reporters: reportersConfigSchema,
}).unknown();
