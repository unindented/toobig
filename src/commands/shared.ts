import { CompositeReporter } from "../reporters";
import { getInputStream, readInputStream } from "../shared";
import {
  Reporter,
  ReporterConfig,
  ReporterConstructor,
  Results,
} from "../types";

export const getCompositeReporter = (
  configReporters: readonly ReporterConfig[]
): Reporter => {
  const reporters = configReporters.map((configReporter) => {
    let reporter: Reporter;

    if (typeof configReporter === "string") {
      const ReporterClass = resolveReporter(configReporter);
      reporter = new ReporterClass();
    } else {
      const ReporterClass = resolveReporter(configReporter[0]);
      reporter = new ReporterClass(configReporter[1]);
    }

    return reporter;
  });

  return new CompositeReporter({ reporters });
};

const resolveReporter = (reporterName: string): ReporterConstructor => {
  let module: { default: ReporterConstructor };

  try {
    module = require(`../reporters/${reporterName}`);
  } catch (err) {
    module = require(reporterName);
  }

  return module.default;
};

export const loadResults = async ({
  results,
}: {
  results: string | Results;
}): Promise<Results> => {
  if (typeof results !== "string") {
    return results;
  }

  const stream = await getInputStream(results);
  const contents = await readInputStream(stream);
  return JSON.parse(contents) as Results;
};

export const reportResults = async ({
  results,
  baselines,
  reporter,
}: {
  results: Results;
  baselines?: Results;
  reporter: Reporter;
}): Promise<void> => {
  await reporter.onRunStart();

  for (const result of Object.values(results)) {
    await reporter.onResult(
      result,
      baselines ? baselines[result.path] : undefined
    );
  }

  await reporter.onRunComplete(results, baselines);
};
