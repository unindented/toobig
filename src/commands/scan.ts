import { resolve as resolvePath } from "path";

import bytes from "bytes";
import glob from "fast-glob";
import getFolderSize from "get-folder-size";

import { scanConfigSchema } from "../schemas";
import { isOverBudget } from "../shared";
import {
  BudgetsConfig,
  Result,
  Results,
  ReturnValue,
  ScanConfig,
  WritableResults,
} from "../types";

import { getCompositeReporter, loadResults, reportResults } from "./shared";

export const scanAndReport = async (
  config: ScanConfig
): Promise<ReturnValue> => {
  await scanConfigSchema.validateAsync(config);

  const {
    budgets,
    cwd = process.cwd(),
    baselines,
    reporters = ["default"],
  } = config;
  const compositeReporter = getCompositeReporter(reporters);

  const scannedResults = await scanResults({ budgets, cwd });
  const baselineResults = baselines
    ? await loadResults({ results: baselines })
    : undefined;
  await reportResults({
    results: scannedResults,
    baselines: baselineResults,
    reporter: compositeReporter,
  });

  const anyOverBudget = Object.values(scannedResults).some(isOverBudget);
  return { results: scannedResults, anyOverBudget };
};

const scanResults = async ({
  budgets,
  cwd,
}: {
  budgets: BudgetsConfig;
  cwd: string;
}): Promise<Results> => {
  const results: WritableResults = {};

  for (const budget of Object.keys(budgets)) {
    const maxSize = bytes(budgets[budget]);
    const paths = await glob(budget, { cwd, onlyFiles: false });

    for (const path of paths) {
      const result: Result = {
        path: path.toString(),
        size: await getSize(resolvePath(cwd, path.toString())),
        maxSize,
      };

      if (results[path] === undefined || results[path].maxSize > maxSize) {
        results[path] = result;
      }
    }
  }

  return results;
};

const getSize = (path: string): Promise<number> =>
  new Promise((resolve, reject) => {
    getFolderSize(path, (err, size) => {
      if (err !== null) {
        reject(err);
        return;
      }
      resolve(size);
    });
  });
