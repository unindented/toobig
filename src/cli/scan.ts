import fromPairs from "lodash/fromPairs";
import { Argv } from "yargs";

import { scanAndReport } from "../commands/scan";
import { ScanConfig } from "../types";

import { configParser } from "./shared";

const pkg = require("../../package.json") as { homepage: string };

export const command = ["scan [options]", "$0"];

export const describe =
  "Scan current working directory to check for entries over budget, and report them";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const builder = (yargs: Argv): Argv<any> =>
  yargs
    .option("budgets", {
      describe: "Budgets to apply when scanning",
      type: "array",
      required: true,
    })
    .coerce("budgets", coerceBudgets)
    .option("cwd", {
      describe: "Current working directory to scan",
      type: "string",
      default: ".",
      normalize: true,
    })
    .option("baselines", {
      describe: "Baseline results file (path or URL)",
      type: "string",
    })
    .option("reporters", {
      describe: "Reporters to process results",
      type: "array",
    })
    .option("config", {
      describe: "Config file, or directory to search for one",
      type: "string",
      default: ".",
      normalize: true,
      config: true,
      configParser,
    })
    .group(
      ["budgets", "cwd", "baselines", "reporters", "config"],
      "Command options:"
    )
    .example("$0 scan", "")
    .example("$0 scan --config myconfig.json", "")
    .example('$0 scan --budgets "dist/*.js" 4KB --budgets "dist/*.css" 2KB', "")
    .example("$0 scan --reporters table summary", "")
    .example("$0 scan --reporters json | $0 load", "")
    .epilogue(`For more details, please see: ${pkg.homepage}`);

export const handler = async (argv: ScanConfig): Promise<boolean> => {
  const { anyOverBudget } = await scanAndReport(argv);
  return anyOverBudget;
};

type ParsedArray = object[] | string[] | string[][];

// istanbul ignore next
const coerceBudgets = (budgets: ParsedArray): object =>
  isNestedStringArray(budgets)
    ? fromPairs(budgets)
    : isStringArray(budgets)
    ? fromPairs([budgets])
    : isObjectArray(budgets)
    ? budgets[0]
    : budgets;

function isNestedStringArray(arr: ParsedArray): arr is string[][] {
  return Array.isArray(arr) && Array.isArray(arr[0]);
}

function isStringArray(arr: ParsedArray): arr is string[] {
  return Array.isArray(arr) && typeof arr[0] === "string";
}

function isObjectArray(arr: ParsedArray): arr is object[] {
  return Array.isArray(arr) && typeof arr[0] === "object";
}
