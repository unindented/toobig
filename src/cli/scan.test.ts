jest.mock("../commands/scan");

import { resolve as resolvePath } from "path";

import yargs, { Arguments } from "yargs";

import { scanAndReport } from "../commands/scan";
import { ScanConfig } from "../types";

import * as scanCommand from "./scan";
import { parserConfiguration } from "./shared";

const pkg = require("../../package.json") as { name: string };

describe("scan", () => {
  let parse: (arg: string) => Promise<string>;

  beforeEach(() => {
    (scanAndReport as jest.Mock).mockResolvedValue({
      results: [],
      anyOverBudget: false,
    });

    const parser = yargs
      .scriptName(pkg.name)
      .command<any>(scanCommand)
      .help()
      .parserConfiguration(parserConfiguration);

    parse = (arg: string): Promise<string> =>
      new Promise((resolve, reject) => {
        parser.parse(
          arg,
          (
            err: Error | undefined,
            _argv: Arguments<ScanConfig>,
            output: string
          ) => {
            if (err) {
              reject(err);
            } else {
              resolve(output);
            }
          }
        );
      });
  });

  describe("with `--help`", () => {
    let output: string;

    beforeEach(async () => {
      output = await parse("scan --help");
    });

    it("shows help output", () => {
      expect(output).toMatchSnapshot();
    });
  });

  describe("with one set of budgets", () => {
    beforeEach(async () => {
      await parse('scan --budgets "dist/*.js" 4KB --cwd dist');
    });

    it("calls `scanAndReport` with the right configuration", () => {
      expect(scanAndReport).toHaveBeenCalledWith(
        expect.objectContaining({
          budgets: {
            "dist/*.js": "4KB",
          },
          cwd: "dist",
        })
      );
    });
  });

  describe("with two sets of budgets", () => {
    beforeEach(async () => {
      await parse('scan --budgets "dist/*.js" 4KB --budgets "dist/*.css" 2KB');
    });

    it("calls `scanAndReport` with the right configuration", () => {
      expect(scanAndReport).toHaveBeenCalledWith(
        expect.objectContaining({
          budgets: {
            "dist/*.js": "4KB",
            "dist/*.css": "2KB",
          },
        })
      );
    });
  });

  describe("with reporters", () => {
    beforeEach(async () => {
      await parse('scan --budgets "dist/*.js" 4KB --reporters table summary');
    });

    it("calls `scanAndReport` with the right configuration", () => {
      expect(scanAndReport).toHaveBeenCalledWith(
        expect.objectContaining({
          reporters: ["table", "summary"],
        })
      );
    });
  });

  describe("with a config file", () => {
    const configPath = resolvePath(
      __dirname,
      "../__fixtures__/scan-config-valid.json"
    );
    const config = require(configPath);

    beforeEach(async () => {
      await parse(`scan --config ${configPath}`);
    });

    it("calls `scanAndReport` with the right configuration", () => {
      expect(scanAndReport).toHaveBeenCalledWith(
        expect.objectContaining(config)
      );
    });
  });
});
