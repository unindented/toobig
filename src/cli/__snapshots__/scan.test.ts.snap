// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scan with \`--help\` shows help output 1`] = `
"toobig scan [options]

Scan current working directory to check for entries over budget, and report them

Command options:
  --budgets    Budgets to apply when scanning                 [array] [required]
  --cwd        Current working directory to scan         [string] [default: \\".\\"]
  --baselines  Baseline results file (path or URL)                      [string]
  --reporters  Reporters to process results                              [array]
  --config     Config file, or directory to search for one
                                                         [string] [default: \\".\\"]

Options:
  --version  Show version number                                       [boolean]
  --help     Show help                                                 [boolean]

Examples:
  toobig scan
  toobig scan --config myconfig.json
  toobig scan --budgets \\"dist/*.js\\" 4KB --budgets \\"dist/*.css\\" 2KB
  toobig scan --reporters table summary
  toobig scan --reporters json | toobig load

For more details, please see: https://github.com/unindented/toobig"
`;
